src/clj/puppetlabs/services/request_handler/request_handler_core.clj

```diff
@@ -12,6 +12,14 @@
             [slingshot.slingshot :as sling]))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Constants
+
+(def error-message-404
+  (str "Error: This master is running Puppet 4, and the URL format has changed. "
+       "This request was likely made with a Puppet 3 Agent, as it used an invalid "
+       "URL format. Please upgrade your agents to Puppet 4 and try again."))
```

Comment:
Huh. Question that just occurred to me: Are we handling 404ed URLs that we *know* correspond to old Puppet versions any differently than we'll handle URLs pointing to Jupiter? Because I'd rather not show this message to someone running, say, Puppet 6.